### PR TITLE
[Elastic Agent] Add support for the --delay-enroll command flag.

### LINKED ISSUE #26667: [Elastic Agent] Auto-enroll on first startup
## Overview
In cloud deployments and/or physical system deployments there is a case where you want an Elastic Agent to enroll into Fleet on first startup and not at installation time. Example being a prepared gold image for a cloud deployment, or a prepared sys-prepped Windows image for an organization.

## Idea
The idea is to add an option to the `install` command to not have it actually start the service and to not actually enroll at that time.

```
.\elastic-agent install --delay-enroll --url https://my.fleet.server --enrollment-token ${token}
```

## Implementation
This would proceed to install the Elastic Agent, write an `enroll.yml` (with the url and enrollment token) file next to the `elastic-agent.yml` file and not start the service.

Then on the first start of the service (which is normally on boot up of a fresh deployment) then the Elastic Agent would notice that a `enroll.yml` file is present, load that perform enrollment, re-exec into Fleet mode, and then delete the `enroll.yml`.

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

This adds support for the `--delay-enroll` flag for the `install` and `enroll` command. On next start of the Elastic Agent service the Elastic Agent will enroll before running in managed mode.

## Why is it important?

This supports the use-case for setting up a golden image, sysprep, or VDI image that will enroll into Fleet on first start.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Closes #26667

### CODE REVIEW

**Comment by ruflin** [comment]
What are the plans on documenting this feature?

File: x-pack/elastic-agent/CHANGELOG.next.asciidoc [hunk 1]
```diff
@@ -124,3 +124,4 @@
 - Enable configuring monitoring namespace {issue}26439[26439]
 - Communicate with Fleet Server over HTTP2. {pull}26474[26474]
 - Support Node and Service autodiscovery in kubernetes dynamic provider. {pull}26801[26801]
+- Add new --enroll-delay option for install and enroll commands. {pull}27118[27118]
```
Comment (ruflin) [inline x-pack/elastic-agent/CHANGELOG.next.asciidoc:127]: @andresrc As we publish this changelog, should we become more verbose here to have more details on what this is about?

**Comment by blakerouse** [comment]
@ruflin What do you mean the plans? I would think it needs to be documented. It is a core use-case we need to support. @nimarezainia can probably provide more on the documentation piece.

**Comment by ruflin** [comment]
@blakerouse Sorry, should have been more specific. Will the docs for this go into observability-docs? Do we already have an issue or PR for it? I'm worried we will forget to document it as it is an other repo.

**Comment by blakerouse** [comment]
@ruflin Filed the documentation request https://github.com/elastic/observability-docs/issues/918

File: x-pack/elastic-agent/pkg/agent/application/paths/files.go [hunk 2]
```diff
@@ -11,27 +11,39 @@ import (
 	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/application/filelock"
 )
 
-// defaultAgentConfigFile is a name of file used to store agent information
+// defaultAgentCapabilitiesFile is a name of file used to store agent capabilities
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/agent/application/paths/files.go:14]: 👍

File: x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go [hunk 3]
```diff
@@ -9,13 +9,16 @@ import (
...
 	"io"
+	"io/ioutil"
 	"math/rand"
 	"net/http"
 	"net/url"
 	"os"
 	"os/exec"
 	"time"
 
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/cli"
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go:20]: please group together

File: x-pack/elastic-agent/pkg/agent/cmd/run.go [hunk 4]
```diff
@@ -296,3 +314,52 @@ func setupMetrics(agentInfo *info.AgentInfo, logger *logger.Logger, operatingSys
 func isProcessStatsEnabled(cfg *monitoringCfg.MonitoringHTTPConfig) bool {
 	return cfg != nil && cfg.Enabled
 }
+
+func tryDelayEnroll(ctx context.Context, logger *logger.Logger, cfg *configuration.Configuration, override cfgOverrider) (*configuration.Configuration, error) {
+	enrollPath := paths.AgentEnrollFile()
+	if _, err := os.Stat(enrollPath); os.IsNotExist(err) {
+		// no enrollment file exists; nothing to do
+		return cfg, nil
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/agent/cmd/run.go:322]: override is not called here, shouldnt this be called on defer maybe

Comment (blakerouse) [reply x-pack/elastic-agent/pkg/agent/cmd/run.go:322]: The `cfg` at that point has already had `overrides` performed on it. `overrides` only needs to be used again in the case that enroll did occur and the entire configuration needs to be read again.

That is why `overrides` is only passed into the `loadConfig` function, and should only be done in the case that the config truly needs to be reloaded.

File: x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go (hunk 3 above)
Comment (blakerouse) [reply x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go]: Fixed!

File: x-pack/elastic-agent/pkg/agent/cmd/run.go (hunk 4 above)
Comment (michalpristas) [reply x-pack/elastic-agent/pkg/agent/cmd/run.go:322]: i see i havent noticed call before

File: x-pack/elastic-agent/pkg/agent/cmd/run.go [hunk 5]
```diff
@@ -296,3 +314,52 @@ func setupMetrics(agentInfo *info.AgentInfo, logger *logger.Logger, operatingSys
 func isProcessStatsEnabled(cfg *monitoringCfg.MonitoringHTTPConfig) bool {
 	return cfg != nil && cfg.Enabled
 }
+
+func tryDelayEnroll(ctx context.Context, logger *logger.Logger, cfg *configuration.Configuration, override cfgOverrider) (*configuration.Configuration, error) {
+	enrollPath := paths.AgentEnrollFile()
+	if _, err := os.Stat(enrollPath); os.IsNotExist(err) {
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/agent/cmd/run.go:320]: err other than FileNotExist is ignored

File: x-pack/elastic-agent/pkg/agent/cmd/run.go [hunk 6]
```diff
@@ -201,6 +190,35 @@ func run(streams *cli.IOStreams, override cfgOverrider) error { // Windows: Mark
...
+			fmt.Sprintf("could not parse configuration file %s", pathConfigFile),
+			errors.TypeFilesystem,
+			errors.M(errors.MetaKeyPath, pathConfigFile))
+	}
+
+	if override != nil {
+		override(cfg)
+	}
+
+	return cfg, err
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/agent/cmd/run.go:219]: small nit: err is nil

**Review by michalpristas — APPROVED** [review]
code LGTM

File: x-pack/elastic-agent/pkg/agent/cmd/run.go (hunk 5 above)
Comment (blakerouse) [reply x-pack/elastic-agent/pkg/agent/cmd/run.go]: True. Updated it to check check for `err != nil`, so any error will cause it to skip the delayed enroll.

File: x-pack/elastic-agent/pkg/agent/cmd/run.go (hunk 6 above)
Comment (blakerouse) [reply x-pack/elastic-agent/pkg/agent/cmd/run.go]: Fixed

**Comment by dikshachauhan-qasource** [comment]
Hi @blakerouse Could you please share steps on how to validate above.

cc @EricDavisX 

Thanks
QAS

**Comment by blakerouse** [comment]
@dikshachauhan-qasource Sure:

Perform the following steps:

1. Run the `install` command just like it is provided in Kibana but add the `--delay-enroll` flag. See that the Elastic Agent is installed, but it is **NOT** running and it is **NOT** showing in Kibana.
2. Restart the machine.
3. Upon restart of the machine you should see a new Elastic Agent (for that machine) appear in Fleet and it should be reporting as healthy.

**Comment by amolnater-qasource** [comment]
Hi @blakerouse 
Thanks for the detailed steps.
We have attempted to test this on 7.15.0 BC-1 and found it working fine.
**Steps followed:**
- We added `--delay-enroll` flag with install command and observe **no agent** was showing in Kibana


- Restarted machine and observed new Agent Under Fleet.


**Build details:**
```
Build: 43636
Commit: d791226d9385122f33f4a5ca38fa5369012fbec3
Artifact Link: https://staging.elastic.co/7.15.0-d9929120/downloads/beats/elastic-agent/elastic-agent-7.15.0-windows-x86_64.zip
```

Thanks
QAS

**Comment by amolnater-qasource** [comment]
Hi @EricDavisX 
We have created **05** testcases for this feature covering Windows, MAC, Linux .tar, .rpm and .deb agents in our Ingest Manager suite at links:
- [C149944](https://elastic.testrail.io/index.php?/cases/view/149944)
- [C149945](https://elastic.testrail.io/index.php?/cases/view/149945)
- [C149946](https://elastic.testrail.io/index.php?/cases/view/149946)
- [C149947](https://elastic.testrail.io/index.php?/cases/view/149947)
- [C149948](https://elastic.testrail.io/index.php?/cases/view/149948)

Please let us know if we are missing anything.
Thanks

**Comment by amolnater-qasource** [comment]
Hi @EricDavisX 
We have executed **05** testcases under [Ingest Manager new Feature Delay Flag 7.15](https://elastic.testrail.io/index.php?/runs/view/2246&group_by=cases:section_id&group_order=asc&group_id=30849) for this feature on 7.15.0 BC-4 Kibana cloud environment.

**Build details:**
Build: 43886
Commit: 29a6969dc230abf16dc65a41c535cb534ae64fa7

Thanks

**Pull request merged**